src/compiler/checker.ts

```diff
@@ -23667,10 +23667,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return false;
     }
 
-    function isTypeParameterAtTopLevel(type: Type, typeParameter: TypeParameter): boolean {
-        return !!(type === typeParameter ||
-            type.flags & TypeFlags.UnionOrIntersection && some((type as UnionOrIntersectionType).types, t => isTypeParameterAtTopLevel(t, typeParameter)) ||
-            type.flags & TypeFlags.Conditional && (getTrueTypeFromConditionalType(type as ConditionalType) === typeParameter || getFalseTypeFromConditionalType(type as ConditionalType) === typeParameter));
+    function isTypeParameterAtTopLevel(type: Type, tp: TypeParameter, depth = 0): boolean {
+        return !!(type === tp ||
+            type.flags & TypeFlags.UnionOrIntersection && some((type as UnionOrIntersectionType).types, t => isTypeParameterAtTopLevel(t, tp, depth)) ||
+            depth < 3 && type.flags & TypeFlags.Conditional && (

```

Comment:
What is the depth check ensuring?

Reply:
That we don't infinitely recurse in conditional types that reference themselves.

Reply:
There are examples of such types in the DT typings for `lodash`.

Reply:
The test added in the PR seems to still pass if I remove `depth < 3`; is there a test we can add which hits this so we don't change this value in a bad way later?

Reply:
Latest commit adds tests up to three levels of conditional types.